File: TypeScript/13.roman-to-integer.ts
/*
 * @lc app=leetcode id=13 lang=typescript
 *
 * [13] Roman to Integer
 */

// @lc code=start
function romanToInt(s: string): number {
    const map = new Map<string, number>([
        ['I', 1],
        ['V', 5],
        ['X', 10],
        ['L', 50],
        ['C', 100],
        ['D', 500],
        ['M', 1000]
    ])
    
    const array: number[] = []
    for (const v of s){
        array.push(map.get(v))
    }

    let result: number = 0
    for (let i = 0 ; i < array.length; i++){
        if (i > 0 && array[i] > array[i-1]) {
            let tmp = array[i] - array[i-1] - array[i-1]
            result += tmp
        }
        else result += array[i]
    }

    return result
};
// @lc code=end

File: Python/13.roman-to-integer.py
#
# @lc app=leetcode id=13 lang=python3
#
# [13] Roman to Integer
#

# @lc code=start
class Solution:
    def romanToInt(self, s: str) -> int:
        d = {'I':1, 'V':5, 'X':10, 'L':50, 'C':100, 'D':500, 'M':1000}
        l = []
        before = 0
        
        for key in s:
            val = d[key]
            
            if before > 0 and val > before:
                l.pop()
                val = (val - before) 
            
            l.append(val)
            before = d[key]
        
        return sum(l)
        
# @lc code=end

